libraries/protocol/include/graphene/protocol/restriction.hpp

```diff
@@ -0,0 +1,129 @@
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#pragma once
+#include <graphene/protocol/base.hpp>
+#include <graphene/protocol/types.hpp>
+
+namespace graphene { namespace protocol {
+
+/**
+    * Defines the set of valid operation restritions as a discriminated union type.
+    */
+struct restriction {
+    enum function_type {
+        func_eq,
+        func_ne,
+        func_lt,
+        func_le,
+        func_gt,
+        func_ge,
+        func_in,
+        func_not_in,
+        func_has_all,
+        func_has_none,
+        func_attr,
+        func_logical_or,
+        FUNCTION_TYPE_COUNT ///< Sentry value which contains the number of different types
+    };
+
+#define GRAPHENE_OP_RESTRICTION_ARGUMENTS_VARIADIC \
+    /*  0 */ void_t, \
+    /*  1 */ bool, \
+    /*  2 */ int64_t, \
+    /*  3 */ string, \
+    /*  4 */ time_point_sec, \
+    /*  5 */ public_key_type, \
+    /*  6 */ fc::sha256, \
+    /*  7 */ account_id_type, \
+    /*  8 */ asset_id_type, \
+    /*  9 */ force_settlement_id_type, \
+    /* 10 */ committee_member_id_type, \
+    /* 11 */ witness_id_type, \
+    /* 12 */ limit_order_id_type, \
+    /* 13 */ call_order_id_type, \
+    /* 14 */ custom_id_type, \
+    /* 15 */ proposal_id_type, \
+    /* 16 */ withdraw_permission_id_type, \
+    /* 17 */ vesting_balance_id_type, \
+    /* 18 */ worker_id_type, \
+    /* 19 */ balance_id_type, \
+    /* 20 */ flat_set<bool>, \
+    /* 21 */ flat_set<int64_t>, \
+    /* 22 */ flat_set<string>, \
+    /* 23 */ flat_set<time_point_sec>, \
+    /* 24 */ flat_set<public_key_type>, \
+    /* 25 */ flat_set<fc::sha256>, \
+    /* 26 */ flat_set<account_id_type>, \
+    /* 27 */ flat_set<asset_id_type>, \
+    /* 28 */ flat_set<force_settlement_id_type>, \
+    /* 29 */ flat_set<committee_member_id_type>, \
+    /* 30 */ flat_set<witness_id_type>, \
+    /* 31 */ flat_set<limit_order_id_type>, \
+    /* 32 */ flat_set<call_order_id_type>, \
+    /* 33 */ flat_set<custom_id_type>, \
+    /* 34 */ flat_set<proposal_id_type>, \
+    /* 35 */ flat_set<withdraw_permission_id_type>, \
+    /* 36 */ flat_set<vesting_balance_id_type>, \
+    /* 37 */ flat_set<worker_id_type>, \
+    /* 38 */ flat_set<balance_id_type>, \
+    /* 39 */ vector<restriction>, \
+    /* 40 */ vector<vector<restriction>>
+
+    using argument_type = fc::static_variant<GRAPHENE_OP_RESTRICTION_ARGUMENTS_VARIADIC>;
+
+    string member_name;
```

Comment:
Good point, I'll work on it

Reply:
So the jsonize to string part is really tricky... If I know the type of the class whose member is named in the {to,from}_variant calls, I can do it... but it's pretty difficult to know that type there, and without that type, there's no context within which those strings can be interpreted.

I started making a `member_index<Class>` type, and that was fine, but I can't put a `member_index<Class>` into a `restriction` without templating `restriction` on `Class`, making each restriction a unique type depending on the object it's restricting, and that just won't fit through our serialization system.

The only way I can think of to do it would be a shim at the API level or something that injects the member name into the JSON as it's heading out on the wire, and restores it on JSON records coming back in before we deserialize them. And that's kind of horrifyingly ugly.

Any better ideas? :\

As to just using `unsigned_int`, that's pretty straightforward and I'm writing it up now.

Reply:
OK done